tabrepo/simulation/ensemble_selection_config_scorer.py

```diff
@@ -135,8 +148,69 @@ def evaluate_task(self, dataset: str, fold: int, models: List[str]) -> Tuple[flo
 
         ensemble_weights: np.array = weighted_ensemble.weights_
 
+        # ensemble_weights has to be updated, need to be in the original models order
+        ensemble_weights_fixed = np.zeros(n_models, dtype=np.float64)
+        ensemble_weights_fixed[models_filtered_idx] = ensemble_weights
+        ensemble_weights = ensemble_weights_fixed
+
         return err, ensemble_weights
 
+    def _get_models_filtered_idx(self, models: list[str], models_filtered: list[str]) -> Tuple[list[str], list[int]]:
+        """
+        Returns the filtered list of models and the index mapping of the filtered models to the original `models` list.
+        """
+        models_filtered_set = set(models_filtered)
+
+        # Preserve `models` order without duplicates
+        seen = set()
+        seen_add = seen.add
+        models_filtered = [m for m in models if (m in models_filtered_set) and not (m in seen or seen_add(m))]
+
+        if len(models_filtered_set) < len(models_filtered):
+            # Duplicate names in `models`, have special handling
+            models_idx = {}
+            for i, m in enumerate(models):
+                if m not in models_idx:
+                    models_idx[m] = []
+                models_idx[m].append(i)
+            models_filtered_idx = [models_idx[m].pop(0) for m in models_filtered]
+        else:
+            models_filtered_idx = [models.index(m) for m in models_filtered]
+        return models_filtered, models_filtered_idx
+
+
+class EnsembleScorerMaxModels(EnsembleScorer):
+    """
+    Identical to EnsembleScorer, with the addition of `max_models` and `max_models_per_type`.
+
+    Parameters
+    ----------
+    max_models: int, default = None
+        If specified, will limit ensemble candidates to the top `max_models` highest validation score models.
+        This logic is applied after the filtering from `max_models_per_type`.
+    max_models_per_type: int, default = None
+        If specified, will limit ensemble candidates of a given model type to the top `max_models_per_type` highest validation score models.
+    """
+    def __init__(self, zeroshot_context: ZeroshotSimulatorContext, max_models: int = None, max_models_per_type: int = None, **kwargs):
+        super().__init__(zeroshot_context=zeroshot_context, **kwargs)
+        assert self.zeroshot_context is not None
+        self.max_models = max_models
```

Comment:
Can we assert that the numbers are both positive? (which I assume is required)

Reply:
Added assert, good call!